store/tikv/latch/latch.go

```diff
@@ -171,29 +211,61 @@ func (latches *Latches) popFromWaitingQueue(slotID int) (front uint64, hasMoreWa
 		delete(latches.waitingQueues, slotID)
 	} else {
 		latches.waitingQueues[slotID] = waiting[1:]
-		hasMoreWaiting = true
+		hasWaiting = true
 	}
 	return
 }
 
-func (latches *Latches) acquireSlot(slotID int, startTS uint64) (success, stale bool) {
+func (latches *Latches) frontOfWaitingQueue(slotID int) *Lock {
+	latches.RLock()
+	defer latches.RUnlock()
+	return latches.waitingQueues[slotID][0]
+}
+
+// acquireSlot tries to acquire a slot first time.
+func (latches *Latches) acquireSlot(slotID int, lock *Lock) {
 	latch := &latches.slots[slotID]
 	latch.Lock()
 	defer latch.Unlock()
-	if stale = latch.maxCommitTS > startTS; stale {
+	if latch.isFree() {
+		latch.acquiredTxn = lock.startTS
+		lock.acquiredCount++
+		lock.status = acquireSuccess
 		return
 	}
-	// Empty latch
-	if !latch.occupied() {
-		latch.waitingQueueHead = startTS
-	}
-	if success = latch.waitingQueueHead == startTS; success {
+	if latch.maxCommitTS > lock.startTS {
+		lock.status = acquireStale
 		return
 	}
-	// push current transaction into waitingQueue
-	latch.hasMoreWaiting = true
+	lock.status = acquireLocked
+	// Push the current transaction into waitingQueue.
+	latch.hasWaiting = true
 	latches.Lock()
 	defer latches.Unlock()
-	latches.waitingQueues[slotID] = append(latches.waitingQueues[slotID], startTS)
-	return
+	latches.waitingQueues[slotID] = append(latches.waitingQueues[slotID], lock)
+}
+
+// retryAcquireSlot retries to acquire slot for current lock.
+// Preconditions: the caller must ensure the last call of acquireSlot
+// is failed.
+func (latches *Latches) retryAcquireSlot(slotID int, lock *Lock) {
+	latch := &latches.slots[slotID]
+	latch.Lock()
+	defer latch.Unlock()
+	if latch.isFree() {
+		panic(fmt.Sprintf("Invalid latch:%#v", latch))
+	}
+	var front *Lock
+	front, latch.hasWaiting = latches.popFromWaitingQueue(slotID)
+	if front.startTS != lock.startTS {
+		panic(fmt.Sprintf("invalid front ts %d, lock:%#v", front.startTS, lock))
+	}
+
+	lock.acquiredCount++
```

Comment:
Check stale first, if it's stale, don't do `lock.acquiredCount++`, and return directly.

Reply:
Then how to determine the number of latches to release when the status is stale? @tiancaiamao